### PR TITLE
Add order service type and extend metadata logic

### PR DESCRIPTION
**Pull request by marcofucci**
This redesigns how we deal with metadata.

1. metadata models can be defined in specific django apps
2. a metadata registry keeps track of all the metadata
3. you need to register a metadata using a `metadata` module with
```python
>>> from datahub.metadata.registry import registry
>>> registry.register(
        id=<metadata-id>,
        model=<model>,
        queryset=<queryset>,
        serializer=<serializer>
    )
```
4. that automatically generates the right view and tests

This also adds a new OMIS ServiceType metadata which uses the new registry logic.

### CODE REVIEW

File: datahub/metadata/test/test_registry.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+def test_register_id_already_used():
+    """
+    Tests that if I try to register the same metadata twice, the second call fails.
+    """
+    reg = MetadataRegistery()
+
+    reg.register('sector', model=Sector)
+
+    with pytest.raises(AssertionError):
```
Comment (reupen) [inline datahub/metadata/test/test_registry.py]: If the behaviour's important, I'd use a proper exception rather an assertion (although we don't use it, assertions are stripped out with `python -O`).

File: datahub/metadata/test/test_views.py [hunk 2]
```diff
@@ -1,115 +1,75 @@
...
+        metafunc.parametrize(
+            'metadata_view_name',
+            view_names,
+            ids=[f'{view_name} metadata view' for view_name in view_names]
+        )
+
+    if 'ordered_mapping' in metafunc.fixturenames:
+        view_data = []
+        for view_id, mapping in registry.mappings.items():
+            with suppress(FieldDoesNotExist):
```
Comment (reupen) [inline datahub/metadata/test/test_views.py]: Would we be able to use `if issubclass(mapping.model, datahub.core.models.BaseOrderedConstantModel):` here?

File: datahub/metadata/test/test_registry.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    """Tests the default values."""
+    reg = MetadataRegistery()
+
+    reg.register('sector', model=Sector)
+    reg.register('uk-region', model=UKRegion)
+
+    assert set(reg.mappings) == {'sector', 'uk-region'}
+
+    mapping = reg.mappings['sector']
+    assert mapping.model == Sector
```
Comment (reupen) [inline datahub/metadata/test/test_registry.py]: Shouldn't that be `mapping.model is Sector`?

File: datahub/metadata/test/test_views.py [hunk 4]
```diff
@@ -118,72 +78,25 @@ def test_metadata_view_patch(view_name, api_client):
...
-
-
-@pytest.mark.parametrize('view_name,expected_names',
-                         ordered_metadata_view_params,
-                         ids=ordered_metadata_view_test_ids)
-def test_ordered_metadata_order_view(view_name, expected_names, api_client):
-    """Test that turnover and no. of employee ranges are returned in order.
-
-    The response elements should be ordered according to the order column in the model.
+    assert set(pattern.name for pattern in patterns) == set(registry.mappings.keys())
```
Comment (reupen) [inline datahub/metadata/test/test_views.py]: Could be simplified a little bit:
`assert {pattern.name for pattern in patterns} == registry.mappings.keys()`

File: datahub/omis/order/serializers.py [hunk 5]
```diff
@@ -3,13 +3,19 @@
...
-from datahub.core.serializers import NestedRelatedField
+from datahub.core.serializers import ConstantModelSerializer, NestedRelatedField
 from datahub.core.validate_utils import DataCombiner
 from datahub.metadata.models import Country, Sector, Team
 
 from .models import Order, OrderAssignee, OrderSubscriber
 
 
+class ServiceTypeSerializer(ConstantModelSerializer):
+    """Service Type DRF serilizer"""
```
Comment (reupen) [inline datahub/omis/order/serializers.py]: Typo in serilizer.

Comment (marcofucci) [inline datahub/omis/order/serializers.py]: 👍

File: datahub/metadata/test/test_views.py (hunk 4 above)
Comment (marcofucci) [inline datahub/metadata/test/test_views.py]: 👍

File: datahub/metadata/test/test_views.py (hunk 2 above)
Comment (marcofucci) [inline datahub/metadata/test/test_views.py]: `issubclass` is not generally pythonic and checking if an object behaves in a certain way is usually preferable.

File: datahub/metadata/test/test_registry.py (hunk 1 above)
Comment (marcofucci) [inline datahub/metadata/test/test_registry.py]: 👍

File: datahub/metadata/test/test_registry.py (hunk 3 above)
Comment (marcofucci) [inline datahub/metadata/test/test_registry.py]: 👍

File: datahub/metadata/registry.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        - serializer (optional): if you want to override the default one
+
+    The data registered is currently used by metadata.views to generate views automatically.
+    """
+
+    def __init__(self):
+        """Keeps a local copy of the metadata registered."""
+        self.metadata = {}
+
+    def register(self, id, model, queryset=None, serializer=ConstantModelSerializer):
```
Comment (marcofucci) [inline datahub/metadata/registry.py]: very true 👍

File: datahub/metadata/registry.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+from collections import namedtuple
+
+from datahub.core.serializers import ConstantModelSerializer
+
+MetadataMapping = namedtuple('MetadataMapping', ['model', 'queryset', 'serializer'])
+
+
+class MetadataRegistery:
```
Comment (marcofucci) [inline datahub/metadata/registry.py]: 👍

File: datahub/metadata/apps.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+class MetadataConfig(AppConfig):
+    """
+    Django App Config for the Metadata app.
+    """
+
+    name = 'datahub.metadata'
+
+    def ready(self):
+        """Calls the autodiscover logic after all apps are loaded."""
+        super(MetadataConfig, self).ready()
```
Comment (marcofucci) [inline datahub/metadata/apps.py]: 👍

**Comment by marcofucci** [comment]
@reupen thanks for the comments, very helpful. Have another look when you can.

**Pull request merged**